server/events/vcs/gitea/client.go

```diff
@@ -0,0 +1,468 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gitea
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"strings"
+	"time"
+
+	"code.gitea.io/sdk/gitea"
+	"github.com/pkg/errors"
+	"github.com/runatlantis/atlantis/server/events/models"
+	"github.com/runatlantis/atlantis/server/logging"
+)
+
+type GiteaClient struct {
+	giteaClient *gitea.Client
+	username    string
+	token       string
+	ctx         context.Context
+	logger      logging.SimpleLogging
+}
+
+type GiteaPRReviewSummary struct {
+	Reviews []GiteaReview
+}
+
+type GiteaReview struct {
+	ID          int64
+	Body        string
+	Reviewer    string
+	State       gitea.ReviewStateType // e.g., "APPROVED", "PENDING", "REQUEST_CHANGES"
+	SubmittedAt time.Time
+}
+
+type GiteaPullGetter interface {
+	GetPullRequest(repo models.Repo, pullNum int) (*gitea.PullRequest, error)
+}
+
+// NewClient builds a client that makes API calls to Gitea. httpClient is the
+// client to use to make the requests, username and password are used as basic
+// auth in the requests, baseURL is the API's baseURL, ex. https://corp.com:7990.
+// Don't include the API version, ex. '/1.0'.
+func NewClient(baseURL string, username string, token string, logger logging.SimpleLogging) (*GiteaClient, error) {
+	giteaClient, err := gitea.NewClient(baseURL,
+		gitea.SetToken(token),
+		gitea.SetUserAgent("atlantis"),
+	)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "creating gitea client")
+	}
+
+	return &GiteaClient{
+		giteaClient: giteaClient,
+		username:    username,
+		token:       token,
+		ctx:         context.Background(),
+		logger:      logger,
+	}, nil
+}
+
+func (c *GiteaClient) GetPullRequest(repo models.Repo, pullNum int) (*gitea.PullRequest, error) {
+	pr, _, err := c.giteaClient.GetPullRequest(repo.Owner, repo.Name, int64(pullNum))
+
+	if err != nil {
+		return nil, err
+	}
+
+	return pr, nil
+}
+
+// GetModifiedFiles returns the names of files that were modified in the merge request
+// relative to the repo root, e.g. parent/child/file.txt.
+func (c *GiteaClient) GetModifiedFiles(repo models.Repo, pull models.PullRequest) ([]string, error) {
+	changedFiles := make([]string, 0)
+	page := 0
+	nextPage := 1
+	listOptions := gitea.ListPullRequestFilesOptions{
+		ListOptions: gitea.ListOptions{
+			Page:     1,
+			PageSize: 100,
+		},
+	}
+
+	for page < nextPage {
+		page = +1
+		listOptions.ListOptions.Page = page
+		files, resp, err := c.giteaClient.ListPullRequestFiles(repo.Owner, repo.Name, int64(pull.Num), listOptions)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, file := range files {
+			changedFiles = append(changedFiles, file.Filename)
+		}
+
+		nextPage = resp.NextPage
+
+		// Emergency break after 500 pages
+		if page >= 500 {
```

Comment:
Why after 500 ?

Reply:
This is a randomly selected value. It **should** never be hit, but is there to prevent a runaway situation. Not sure if this should be a constant or a configurable value honestly.

Reply:
Let's start with a constant

Reply:
Change into const with some comments next to it.